clients/src/main/java/org/apache/kafka/clients/NetworkClientUtils.java

```diff
@@ -88,8 +89,23 @@ public static boolean awaitReady(KafkaClient client, Node node, Time time, long
      * care.
      */
     public static ClientResponse sendAndReceive(KafkaClient client, ClientRequest request, Time time) throws IOException {
+        return sendAndReceive(client, request, time, () -> false);
+    }
+
+    /**
+     * Invokes `client.send` followed by 1 or more `client.poll` invocations until a response is received or a
+     * disconnection happens (which can happen for a number of reasons including a request timeout). The polling
+     * loop may be stopped from another thread by setting `shutdownInitiated` flag.
+     *
+     * In case of a disconnection, an `IOException` is thrown.
+     *
+     * This method is useful for implementing blocking behaviour on top of the non-blocking `NetworkClient`, use it with
+     * care.
+     */
+    public static ClientResponse sendAndReceive(KafkaClient client, ClientRequest request, Time time,
+                                                Supplier<Boolean> shutdownInitiated) throws IOException {
         client.send(request, time.milliseconds());
-        while (true) {
+        while (!shutdownInitiated.get()) {
```

Comment:
Wouldn't it be better to have a way to initiate an orderly shutdown of `KafkaClient`? It seems a bit more general.

Reply:
@ijuma Thanks for the review. I have updated with a new method in KafkaClient to initiate shutdown. Not sure if it matches what you had in mind. Let me know what you think.